cache.go

```diff
@@ -1,91 +1,266 @@
 	return Stats{
-		Hits:   atomic.LoadInt64(&s.Hits) - o.Hits,
-		Misses: atomic.LoadInt64(&s.Misses) - o.Misses,
+		Count:     stats.Count,
+		Sets:      stats.Sets - previous.Sets,
+		Gets:      stats.Gets - previous.Gets,
+		Hits:      stats.Hits - previous.Hits,
+		Misses:    stats.Misses - previous.Misses,
+		Evictions: stats.Evictions - previous.Evictions,
 	}
 }
 
-type entry struct {
+// Configuration for the Cache.
+type Config struct {
+	// Maximum number of items in the cache
+	Size int
+	// Optional time before an item expires
+	MaxAge time.Duration
+	// Optional callback invoked when an item is evicted due to the LRU policy
+	OnEviction func(key, value interface{})
+	// Optional callback invoked when an item expired
+	OnExpiration func(key, value interface{})
+}
+
+// Entry pointed to by each list.Element
+type cacheEntry struct {
+	key     interface{}
 	value   interface{}
 	created time.Time
 }
 
 // LRU implements a thread-safe fixed size LRU cache.
 type Cache struct {
-	lru    *simplelru.LRU
-	mu     sync.Mutex
-	maxAge time.Duration
-	stats  Stats
-}
-
-// New constructs an LRU of the given size and max age to return
-// results for
-func New(size int, maxAge time.Duration) (*Cache, error) {
-	l, err := simplelru.NewLRU(size, nil)
-	if err != nil {
-		return nil, err
+	// Fields defined by configuration
+	size         int
+	maxAge       time.Duration
+	onEviction   func(key, value interface{})
+	onExpiration func(key, value interface{})
+
+	// Cache statistics
+	sets      int64
+	gets      int64
+	hits      int64
+	misses    int64
+	evictions int64
+
+	items        map[interface{}]*list.Element
+	evictionList *list.List
+	mutex        sync.RWMutex
+}
+
+// New constructs an LRU Cache with the given Config object. Returns an error
+// given an invalid size.
+func New(config Config) (*Cache, error) {
+	if config.Size <= 0 {
+		return nil, errors.New("Must supply a positive config.Size")
 	}
 
-	c := &Cache{
-		lru:    l,
-		maxAge: maxAge,
+	if config.MaxAge < 0 {
+		return nil, errors.New("Must supply a zero or positive config.MaxAge")
 	}
 
-	return c, nil
+	cache := &Cache{
+		size:         config.Size,
+		maxAge:       config.MaxAge,
+		onEviction:   config.OnEviction,
+		onExpiration: config.OnExpiration,
+		items:        make(map[interface{}]*list.Element),
+		evictionList: list.New(),
+	}
+
+	return cache, nil
 }
 
-// Add adds an additional key/value pair to our cache.
-func (c *Cache) Add(key, value interface{}) bool {
-	c.mu.Lock()
-	defer c.mu.Unlock()
+// Set updates a key:value pair in the cache. Returns true if an eviction
+// occurrred, and subsequently invokes the OnEviction callback.
+func (cache *Cache) Set(key, value interface{}) bool {
+	cache.mutex.Lock()
+	defer cache.mutex.Unlock()
 
-	return c.lru.Add(key, entry{
-		value:   value,
-		created: time.Now(),
-	})
+	cache.sets++
+	created := time.Now()
+
+	if element, ok := cache.items[key]; ok {
+		cache.evictionList.MoveToFront(element)
```

Comment:
should we increment `hits` too?

Reply:
I think we only want to increment hits/misses on Get operations. Consider the following:
``` go
for i := 0; i <= 9; i++ {
	cache.Set("foo", i)
}
a, _ := cache.Get("foo")
v, _ := cache.Get("bar")
```
If we increment hits during Set, then the result is:
Hits: 10
Misses: 1

If we only increment during Get, then:
Hits: 1
Misses: 1

I think the latter is a better representation of what we consider a "hit" in caching: we have a 50% hit rate when performing a Get, not 1/11 out of both Get/Set.

Reply:
ah gotcha, maybe we shouldn't promote a key on `.Set()` at all then?

Reply:
@yields Imagine two `Set()` operations in a row, the last of which would exceed capacity.

* If we were never to promote the key on `.Set()`, but instead push it to the end of the list... Rather than remove a key which hasn't been accessed in a while, we're deleting a key that was set just moments earlier. We have essentially no data on the access frequency for that key.
* If we promote the key on `.Set()`, then we're removing a key for which we have some understanding of its access given a read-heavy cache

As a note, this behavior is consistent with the previous implementation based on github.com/hashicorp/golang-lru/simplelru

As well as literature and implementations I've found elsewhere:
* https://blogs.dropbox.com/tech/2012/10/caching-in-theory-and-practice/
* http://openmymind.net/Writing-An-LRU-Cache/
* https://chrisrng.svbtle.com/lru-cache-in-javascript
* https://alaindefrance.wordpress.com/2014/10/05/lru-cache-implementation/
* https://github.com/isaacs/node-lru-cache

Reply:
Maybe what should be clarified in documentation is that Hits/Misses pertain to fetches only?